### PR TITLE
Documentation: intro and configuration

### PR DESCRIPTION
**Pull request by tristantarrant**
- split out the introduction into a dedicated chapter
- align some statements about the configuration with the latest status quo
- describe how configuration templates work

### CODE REVIEW

File: documentation/src/main/asciidoc/user_guide/configuration.adoc [hunk 1]
```diff
@@ -131,14 +200,13 @@ Cache<String, String> cache = manager.getCache(newCacheName);
...
 
 ==== ConfigurationBuilder Programmatic Configuration API
-However, users do not have to first read an XML based configuration and then modify it in runtime; they can start from scratch using only programmatic API.
-This is where powerful ConfigurationBuilder API comes to shine.
-The aim of this API is to make it easier to chain coding of configuration options in order to speed up the coding itself and make the configuration more readable.
-This new configuration can be used for both the global and the cache level configuration.
+While the above paragraph shows how to combine declarative and programmatic configuration, starting from an XML configuration is completely optional.
+The ConfigurationBuilder API's fluent approach is to make it easier to chain coding of configuration options in order to speed up the coding itself and make the configuration more readable.
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: I think something simpler along the lines
```
The ConfigurationBuilder fluent interface style allows for easier to write and more readable programmatic configuration.
``` 
would be more easy on the eyes

File: documentation/src/main/asciidoc/user_guide/configuration.adoc [hunk 2]
```diff
@@ -276,51 +344,35 @@ Configuration config = new ConfigurationBuilder()
...
-Infinispan has a number of scripts for importing configurations from other cache and data grid products.
-Currently we have scripts to import configurations from:
-
-* JBoss Cache 3.x
-* EHCache 1.x
-* Oracle Coherence 3.x
+The configuration format of Infinispan has changed since version 6.0 in order to align the embedded schema with the one used
+by the server. For this reason, when upgrading to Infinispan 7.x or later, you should use the configuration converter included in the
+_all_ distribution. Simply invoke it from the command-line passing the old configuration file as the first parameter and the name
+of the converted file as the second parameter. The converted On Unix/Linux/macOS:
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: s/The converted On/To convert on

File: documentation/src/main/asciidoc/user_guide/configuration.adoc [hunk 3]
```diff
@@ -350,17 +402,17 @@ and if you happen to use an XML file to configure Infinispan, just use:
...
 
 The configurations available are:
 
-*  default-jgroups-udp.xml - Uses UDP as a transport, and UDP multicast for discovery.  Usually suitable for larger (over 100 nodes) clusters _or_ if you are using link:http://community.jboss.org/docs/DOC-14853#replicated[replication or invalidation] .  Minimises opening too many sockets. 
-*  default-jgroups-tcp.xml - Uses TCP as a transport and UDP multicast for discovery.  Better for smaller clusters (under 100 nodes) _only if_ you are using link:http://community.jboss.org/docs/DOC-14853#distribution[distribution] , as TCP is more efficient as a point-to-point protocol 
-*  default-jgroups-ec2.xml - Uses TCP as a transport and link:http://community.jboss.org/docs/DOC-15925[S3_PING] for discovery.  Suitable on link:http://aws.amazon.com/ec2/[Amazon EC2] nodes where UDP multicast isn't available. 
+*  default-jgroups-udp.xml - Uses UDP as a transport, and UDP multicast for discovery.  Usually suitable for larger (over 100 nodes) clusters _or_ if you are using link:http://community.jboss.org/docs/DOC-14853#replicated[replication or invalidation] .  Minimises opening too many sockets.
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: Outdated link

Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: Extra spaces before ```Minimises```

File: documentation/src/main/asciidoc/user_guide/configuration.adoc [hunk 4]
```diff
@@ -350,17 +402,17 @@ and if you happen to use an XML file to configure Infinispan, just use:
...
+*  default-jgroups-tcp.xml - Uses TCP as a transport and UDP multicast for discovery.  Better for smaller clusters (under 100 nodes) _only if_ you are using link:http://community.jboss.org/docs/DOC-14853#distribution[distribution] , as TCP is more efficient as a point-to-point protocol
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: same as above, link + spaces

File: documentation/src/main/asciidoc/user_guide/configuration.adoc [hunk 5]
```diff
@@ -350,17 +402,17 @@ and if you happen to use an XML file to configure Infinispan, just use:
...
+*  default-jgroups-ec2.xml - Uses TCP as a transport and link:http://community.jboss.org/docs/DOC-15925[S3_PING] for discovery.  Suitable on link:http://aws.amazon.com/ec2/[Amazon EC2] nodes where UDP multicast isn't available.
 *  default-jgroups-kubernetes.xml - Uses TCP as a transport and link:https://github.com/jgroups-extras/jgroups-kubernetes[KUBE_PING] for discovery.  Suitable on link:http://kubernetes.io/[Kubernetes] and link:https://www.openshift.org/[OpenShift] nodes where UDP multicast is not always available.
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: idem

File: documentation/src/main/asciidoc/user_guide/configuration.adoc [hunk 6]
```diff
@@ -408,7 +460,7 @@ $ java -cp ... -Djgroups.tcp.port=1234 -Djgroups.tcp.address=10.11.12.13
 
 
 ==== Further reading
-JGroups also supports more system property overrides, details of which can be found on this page: link:http://community.jboss.org/docs/12352[SystemProps] 
+JGroups also supports more system property overrides, details of which can be found on this page: link:http://community.jboss.org/docs/12352[SystemProps]
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: Is there a more recent link to this? That one ^ is a 10 year old page updated last 6 years ago  :smile:

File: documentation/src/main/asciidoc/user_guide/configuration.adoc [hunk 7]
```diff
@@ -58,23 +54,96 @@ Cache defaultCache = manager.getCache();
...
+   </cache-container>
+</infinispan>
+
+----
+
+In the above example, `base-template` defines a local cache with a specific _expiration_ configuration. The `extended-template`
+configuration inherits from `base-template`, overriding just a single parameter of the _expiration_ element (all other
+attributes are inherited) and adds an _eviction_ element. Finally, two caches are defined: `local` which uses the `base-template`
+configuration and `local-bounded` which uses the `extended-template` configuration.
+
```
Comment (tristantarrant) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: I believe the element is completely overridden but I need to check

File: documentation/src/main/asciidoc/user_guide/configuration.adoc (hunk 3 above)
Comment (tristantarrant) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: I'll replace those links with anchors in the user guide. Odd that this was missed in the first conversion from the wiki

File: documentation/src/main/asciidoc/user_guide/configuration.adoc (hunk 7 above)
Comment (tristantarrant) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: I actually found a bug in this behaviour, so I've opened https://github.com/infinispan/infinispan/pull/4591

File: documentation/src/main/asciidoc/user_guide/configuration.adoc (hunk 6 above)
Comment (tristantarrant) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: I've asked @belaban to find a better home for that and he has opened https://issues.jboss.org/browse/JGRP-2107
We will update this when it becomes available

File: documentation/src/main/asciidoc/user_guide/configuration.adoc (hunk 7 above)
Comment (tristantarrant) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: The behaviour is to merge, and I'll add a paragraph warning about this

Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: right, just to be clear here's the use case I had in mind:

```xml
<cache configuration="my-config">
   <indexing index="LOCAL">
      <property name="prop1">value1</property>
   </indexing>
</cache>

<cache name="mycache" configuration="my-config">
   <indexing index="LOCAL">
      <property name="prop2">value1</property>
   </indexing>
</cache>
```

**Comment by tristantarrant** [comment]
Addressed all of the comments and added some "why use Infinispan" items to the introduction

File: documentation/src/main/asciidoc/user_guide/configuration.adoc (hunk 6 above)
Comment (tristantarrant) [inline documentation/src/main/asciidoc/user_guide/configuration.adoc]: I've updated it with the link to the new JGroups manual.

**Comment by tristantarrant** [comment]
I have made the changes you asked @wburns

**Comment by wburns** [comment]
Integrated, thanks @tristantarrant !

**Pull request merged**